File: ShapeSystem/ShapeSystem.Domain/Shapes.cs
// FILE: ShapeSystem.Domain/Shapes.cs
// LAYER: Domain — pure math, zero I/O

using System;

namespace ShapeSystem.Domain
{
    // The abstract base that every shape must inherit from.
    // "abstract" means you can never do: new Shape() — only new Circle(), etc.
    public abstract class Shape
    {
        public abstract string ShapeType { get; }
        public abstract double CalculateArea();

        public override string ToString() =>
            $"{ShapeType}: area = {CalculateArea():F4}";
    }

    public class Circle : Shape
    {
        public double Radius { get; }

        public Circle(double radius)
        {
            // Protect the invariant: radius can never be zero or negative
            if (radius <= 0)
                throw new ArgumentException("Radius must be greater than zero.", nameof(radius));
            Radius = radius;
        }

        public override string ShapeType => "Circle";
        public override double CalculateArea() => Math.PI * Radius * Radius;
    }

    public class Rectangle : Shape
    {
        public double Width { get; }
        public double Height { get; }

        public Rectangle(double width, double height)
        {
            if (width <= 0) throw new ArgumentException("Width must be > 0.", nameof(width));
            if (height <= 0) throw new ArgumentException("Height must be > 0.", nameof(height));
            Width = width;
            Height = height;
        }

        public override string ShapeType => "Rectangle";
        public override double CalculateArea() => Width * Height;
    }

    public class Triangle : Shape
    {
        public double Base { get; }
        public double Height { get; }

        public Triangle(double @base, double height)
        {
            if (@base <= 0) throw new ArgumentException("Base must be > 0.", nameof(@base));
            if (height <= 0) throw new ArgumentException("Height must be > 0.", nameof(height));
            Base = @base;
            Height = height;
        }

        public override string ShapeType => "Triangle";
        public override double CalculateArea() => 0.5 * Base * Height;
    }

    // NEW SHAPE = just add a new class here. Nothing else in the project changes.
    public class Hexagon : Shape
    {
        public double SideLength { get; }

        public Hexagon(double sideLength)
        {
            if (sideLength <= 0)
                throw new ArgumentException("Side length must be > 0.", nameof(sideLength));
            SideLength = sideLength;
        }

        public override string ShapeType => "Hexagon";
        public override double CalculateArea() =>
            (3 * Math.Sqrt(3) / 2) * SideLength * SideLength;
    }
}

File: js language/shapes.js
// FILE: shape-system/shapes.js
// LAYER: Domain — pure math. No console.log. No document. No require('./loggers').

'use strict';

/**
 * Abstract base class for all shapes.
 * Calling new Shape() directly will throw.
 */
class Shape {
  get shapeType() {
    throw new Error(`${this.constructor.name} must implement shapeType getter.`);
  }

  calculateArea() {
    throw new Error(`${this.constructor.name} must implement calculateArea().`);
  }

  toString() {
    return `${this.shapeType}: area = ${this.calculateArea().toFixed(4)}`;
  }
}

class Circle extends Shape {
  /** @param {number} radius */
  constructor(radius) {
    super();
    // Protect the invariant: negative radius is nonsense
    if (typeof radius !== 'number' || radius <= 0) {
      throw new RangeError(`Radius must be a positive number. Got: ${radius}`);
    }
    this._radius = radius;  // _radius = private by convention
  }

  get shapeType() { return 'Circle'; }
  get radius()    { return this._radius; }  // read-only access

  calculateArea() {
    return Math.PI * this._radius ** 2;
  }
}

class Rectangle extends Shape {
  /** @param {number} width @param {number} height */
  constructor(width, height) {
    super();
    if (typeof width  !== 'number' || width  <= 0) throw new RangeError(`Width must be > 0. Got: ${width}`);
    if (typeof height !== 'number' || height <= 0) throw new RangeError(`Height must be > 0. Got: ${height}`);
    this._width  = width;
    this._height = height;
  }

  get shapeType() { return 'Rectangle'; }
  calculateArea() { return this._width * this._height; }
}

class Triangle extends Shape {
  /** @param {number} base @param {number} height */
  constructor(base, height) {
    super();
    if (typeof base   !== 'number' || base   <= 0) throw new RangeError(`Base must be > 0. Got: ${base}`);
    if (typeof height !== 'number' || height <= 0) throw new RangeError(`Height must be > 0. Got: ${height}`);
    this._base   = base;
    this._height = height;
  }

  get shapeType() { return 'Triangle'; }
  calculateArea() { return 0.5 * this._base * this._height; }
}

// ADDING A NEW SHAPE = add a class here. Nothing else in the project changes.
class Hexagon extends Shape {
  /** @param {number} sideLength */
  constructor(sideLength) {
    super();
    if (typeof sideLength !== 'number' || sideLength <= 0) {
      throw new RangeError(`Side length must be > 0. Got: ${sideLength}`);
    }
    this._sideLength = sideLength;
  }

  get shapeType() { return 'Hexagon'; }
  calculateArea() {
    return (3 * Math.sqrt(3) / 2) * this._sideLength ** 2;
  }
}

// Export all classes so other files can import them
module.exports = { Shape, Circle, Rectangle, Triangle, Hexagon };